File: python/privacy_kit/retention.py
"""Retención y borrado automático. Cumple el principio de plazo y el derecho al olvido."""
from __future__ import annotations
from datetime import datetime, timezone, timedelta

from .config import PrivacyConfig
from .store.base import PrivacyStore


class RetentionManager:
    def __init__(self, store: PrivacyStore, config: PrivacyConfig):
        self.store = store
        self.config = config

    def expires_at(self, category: str, created_iso: str | None = None) -> str | None:
        days = self.config.retention(category)
        if days <= 0:
            return None
        base = datetime.fromisoformat(created_iso) if created_iso else datetime.now(timezone.utc)
        return (base + timedelta(days=days)).isoformat()

    def sweep(self, collections: list[str]) -> int:
        """Borra registros vencidos. Agendar en cron/worker (ej. diario)."""
        now = datetime.now(timezone.utc).isoformat()
        total = 0
        for col in collections:
            for row in self.store.expired(col, now):
                total += self.store.delete(col, {"id": row.get("id")})
        return total

File: node/src/retention.js
'use strict';
// Retención y borrado automático (principio de plazo + derecho al olvido).
class RetentionManager {
  constructor(store, config) { this.store = store; this.config = config; }

  expiresAt(category, createdIso = null) {
    const days = this.config.retention(category);
    if (days <= 0) return null;
    const base = createdIso ? new Date(createdIso) : new Date();
    return new Date(base.getTime() + days * 86400000).toISOString();
  }

  // Borra registros vencidos. Agendar en cron/worker (ej. diario).
  sweep(collections) {
    const now = new Date().toISOString();
    let total = 0;
    for (const col of collections)
      for (const row of this.store.expired(col, now)) total += this.store.delete(col, { id: row.id });
    return total;
  }
}
module.exports = { RetentionManager };
